drivers/adc/adc_mcp3914.c

```diff
@@ -0,0 +1,847 @@
+			   BIT(config->interrupt.pin));
+
+	if (gpio_add_callback(config->interrupt.port, &(data->callback_data_ready))) {
+		LOG_ERR("%s: Failed to add data ready callback", dev->name);
+		return -EINVAL;
+	}
+
+	gpio_pin_interrupt_configure_dt(&config->interrupt, GPIO_INT_EDGE_TO_INACTIVE);
+
+	k_tid_t tid = k_thread_create(
+		&data->thread, data->stack, CONFIG_ADC_MCP3914_ACQUISITION_THREAD_STACK_SIZE,
+		(k_thread_entry_t)mcp3914_acquisition_thread, data, NULL, NULL,
+		CONFIG_ADC_MCP3914_ACQUISITION_THREAD_PRIO, 0, K_FOREVER);
+
+	if (!tid) {
+		LOG_ERR("%s cannot create acq thread", dev->name);
+		return -ENOSYS;
+	}
+
+	if (IS_ENABLED(CONFIG_THREAD_NAME)) {
+		k_thread_name_set(tid, dev->name);
+	}
+
+	k_thread_start(tid);
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	LOG_INF("MCP3941 device (%s) initialised", dev->name);
+
+	return 0;
+}
+
+static const struct adc_driver_api mcp3914_adc_api = {
+	.channel_setup = mcp3914_channel_setup,
+	.read = mcp3914_read,
+	.ref_internal = (uint16_t)MCP3914_INTERNAL_VOLTAGE_REFERENCE,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = mcp3914_read_async,
+#endif
+};
+
+#define MCP3491_CONF_IRQ(inst) GPIO_DT_SPEC_INST_GET(inst, dready_gpios)
+
+#define MCP3491_ADC_INIT(inst)                                                                     \
+                                                                                                   \
+	static const struct mcp3914_config mcp3914_config_##inst = {                               \
+		.bus = SPI_DT_SPEC_GET(DT_DRV_INST(inst),                                          \
+				       SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB | SPI_WORD_SET(8),    \
+				       0),                                                         \
+		.interrupt = COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, dready_gpios),                \
+					 (MCP3491_CONF_IRQ(inst)), ({0})),                         \
+		.channels = DT_PROP(DT_DRV_INST(inst), channel_count),                             \
+		.presc = DT_STRING_UNQUOTED(DT_DRV_INST(inst), amclk_prescale),                    \
+		.osr = DT_STRING_UNQUOTED(DT_DRV_INST(inst), hware_osr),                           \
+		.clksrc = DT_STRING_UNQUOTED(DT_DRV_INST(inst), mclk_src)};                        \
+                                                                                                   \
+	static struct mcp3914_data mcp3914_data_##inst = {                                         \
+		ADC_CONTEXT_INIT_LOCK(mcp3914_data_##inst, ctx),                                   \
+		ADC_CONTEXT_INIT_SYNC(mcp3914_data_##inst, ctx),                                   \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, &mcp3914_init, NULL, &mcp3914_data_##inst,                     \
+			      &mcp3914_config_##inst, POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,       \
```

Comment:
Should not use same as SPI_INIT_PRIORITY as SPI needs to be initialized before MCP driver. Maybe make it configuration